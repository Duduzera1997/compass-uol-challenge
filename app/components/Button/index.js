import React from 'react';
import PropTypes from 'prop-types';
import * as SC from './styles';

function Button({ onPress }) {
  return (
    <SC.Container>
      <SC.TouchableButton onPress={onPress} testID="component-button-press">
        <SC.Label>Visualizar Not√≠cia Completa</SC.Label>
      </SC.TouchableButton>
    </SC.Container>
  );
}

Button.propTypes = {
  onPress: PropTypes.func.isRequired,
};

export default Button;
