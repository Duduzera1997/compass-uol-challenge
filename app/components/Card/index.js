import React from 'react';
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import ptBR from 'date-fns/locale/pt-BR';
import * as SC from './styles';

function Card({ imageUrl, title, abstract, noticeDate }) {
  return (
    <SC.ContainerCard>
      <SC.CardTouchable>
        <SC.CardContent flex={1}>
          <SC.PreviewImage source={{ uri: imageUrl }} />
        </SC.CardContent>
        <SC.CardContent flex={2.5}>
          <SC.Text size={15} fontBold>
            {title}
          </SC.Text>
          <SC.TextDate size={11.5} fontBold>
            {format(new Date(noticeDate), "dd 'de' MMMM 'de' yyyy", {
              locale: ptBR,
            })}
          </SC.TextDate>
          <SC.Text numberOfLines={2} size={13}>
            {abstract}
          </SC.Text>
        </SC.CardContent>
      </SC.CardTouchable>
    </SC.ContainerCard>
  );
}

Card.propTypes = {
  imageUrl: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  abstract: PropTypes.string.isRequired,
  noticeDate: PropTypes.string.isRequired,
};

export default Card;
