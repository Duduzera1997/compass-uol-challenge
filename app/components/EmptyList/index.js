import React from 'react';
import PropTypes from 'prop-types';
import * as SC from './styles';

function EmptyList({ isLoading }) {
  return (
    <SC.ContainerEmptyList>
      {isLoading && <SC.ActivityIndicator />}
      <SC.Label>
        {isLoading
          ? 'Buscando...'
          : 'Ooops! ðŸ˜ž\n Nada foi encontrado, tente novamente...'}
      </SC.Label>
    </SC.ContainerEmptyList>
  );
}

EmptyList.defaultProps = {
  isLoading: false,
};

EmptyList.propTypes = {
  isLoading: PropTypes.bool,
};

export default EmptyList;
