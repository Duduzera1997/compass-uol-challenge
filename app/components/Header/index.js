import PropTypes from 'prop-types';
import React from 'react';

import * as SC from './styles';
import { formatDateLiteral } from '~/utils';

function Header({ headerText, dateText }) {
  return (
    <SC.ContainerHeader>
      <SC.HeaderTitle>{headerText} News</SC.HeaderTitle>
      <SC.DateLabel>Hoje Ã©, {dateText}.</SC.DateLabel>
    </SC.ContainerHeader>
  );
}

Header.defaultProps = {
  dateText: formatDateLiteral(new Date()),
};

Header.propTypes = {
  headerText: PropTypes.string.isRequired,
  dateText: PropTypes.string,
};

export default Header;
