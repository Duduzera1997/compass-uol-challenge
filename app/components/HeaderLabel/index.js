import React from 'react';
import { View, Text } from 'react-native';
import PropTypes from 'prop-types';

function HeaderLabel({ label }) {
  return (
    <View>
      <Text>{label} News</Text>
    </View>
  );
}

HeaderLabel.defaultProps = {
  label: 'Best',
};

HeaderLabel.propTypes = {
  label: PropTypes.string,
};

export default HeaderLabel;
