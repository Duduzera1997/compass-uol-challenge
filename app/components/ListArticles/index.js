import React, { useState, useEffect } from 'react';
import { RefreshControl } from 'react-native';
import PropTypes from 'prop-types';
import _ from 'lodash';

import * as SC from './styles';
import ArticlesDataPropTypes from './defaultPropTypes';
import Card from '../Card';
import EmptyList from '../EmptyList';
import { PAGE_LIMIT } from '~/utils/constants';

function ListArticles({ data, loading, onRefresh, onCardPress }) {
  const [articles, setArticles] = useState({ items: [], currentPage: 1 });

  useEffect(() => {
    if (articles.items.length !== data.results.length) {
      setArticles({
        ...articles,
        items: data?.results ? data.results.slice(0, PAGE_LIMIT) : [],
        currentPage: 1,
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data]);

  function getOffSet(page) {
    return (page - 1) * PAGE_LIMIT;
  }

  function getFakePagination() {
    const { currentPage, items } = articles;
    if (Math.ceil(data.num_results / PAGE_LIMIT) <= currentPage) return;
    const nextPage = currentPage + 1;
    const offset = getOffSet(nextPage);
    setArticles({
      items: [...items, ...data.results.slice(offset, PAGE_LIMIT + offset)],
      currentPage: nextPage,
    });
  }

  function getItem(list, index) {
    return list[index];
  }

  function getItemCount(list) {
    return list.length;
  }

  function renderItem({ item }) {
    const { multimedia, title, abstract, published_date } = item;
    return (
      multimedia && (
        <Card
          imageUrl={multimedia[multimedia.length - 1].url}
          title={title}
          abstract={abstract}
          noticeDate={published_date}
          onCardPress={() => onCardPress(item)}
        />
      )
    );
  }

  return (
    <SC.VirtualList
      data={articles.items}
      refreshControl={
        <RefreshControl
          enabled={!loading && !!data.num_results}
          refreshing={loading}
          onRefresh={onRefresh}
        />
      }
      renderItem={renderItem}
      keyExtractor={(item) => item.url}
      getItemCount={getItemCount}
      getItem={getItem}
      onEndReachedThreshold={0.1}
      keyboardShouldPersistTaps="never"
      onEndReached={getFakePagination}
      ListEmptyComponent={<EmptyList isLoading={loading} />}
    />
  );
}

ListArticles.propTypes = {
  data: ArticlesDataPropTypes.isRequired,
  loading: PropTypes.bool.isRequired,
  onRefresh: PropTypes.func.isRequired,
  onCardPress: PropTypes.func.isRequired,
};

export default ListArticles;
