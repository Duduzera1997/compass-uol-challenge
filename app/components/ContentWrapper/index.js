import React from 'react';
import PropTypes from 'prop-types';
import * as SC from './styles';

import Header from '../Header';
import LinearBackground from '../LinearBackground';
import SearchInput from '../SearchInput';
import ViewArticleModal from '../ViewArticleModal';

function ContentWrapper({
  title,
  children,
  showSearchInput,
  onSeachInputTextChange,
  modalProps: { visible, item, onRequestClose },
}) {
  return (
    <LinearBackground>
      <SC.Container>
        <ViewArticleModal
          visible={visible}
          item={item}
          onRequestClose={onRequestClose}
        />
        <SC.ContainerByFlex flex={1}>
          <SC.ContainerByFlex flex={1}>
            <Header headerText={title} />
          </SC.ContainerByFlex>
          {showSearchInput && (
            <SC.ContainerByFlex flex={1}>
              <SearchInput onChangeText={onSeachInputTextChange} />
            </SC.ContainerByFlex>
          )}
          <SC.ContainerByFlex flex={6}>{children}</SC.ContainerByFlex>
        </SC.ContainerByFlex>
      </SC.Container>
    </LinearBackground>
  );
}

ContentWrapper.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  showSearchInput: PropTypes.bool.isRequired,
  onSeachInputTextChange: PropTypes.func.isRequired,
  modalProps: PropTypes.shape({
    visible: PropTypes.bool.isRequired,
    item: PropTypes.shape({}),
    onRequestClose: PropTypes.func,
  }).isRequired,
};

export default ContentWrapper;
