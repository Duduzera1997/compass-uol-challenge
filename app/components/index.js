import TabIcon from './TabIcon';
import LinearBackground from './LinearBackground';
import Loading from './Loading';
import Header from './Header';
import Card from './Card';
import ListHelper from './ListHelper';
import ListArticles from './ListArticles';
import ViewArticleModal from './ViewArticleModal';
import Button from './Button';
import SearchInput from './SearchInput';
import ContentWrapper from './ContentWrapper';

export {
  TabIcon,
  LinearBackground,
  Loading,
  Header,
  Card,
  ListHelper,
  ListArticles,
  ViewArticleModal,
  Button,
  SearchInput,
  ContentWrapper,
};
