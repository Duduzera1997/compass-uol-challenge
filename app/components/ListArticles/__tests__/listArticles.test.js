import React from 'react';
import { ListArticles } from '~/components';
import { renderWithTheme } from '~/utils/helperTest';

const mock = {
  section: 'Science',
  last_updated: '2021-03-19T17:10:56-04:00',
  num_results: 1,
  results: [
    {
      title: 'How to Vaccinate Homebound Seniors? Take the Shots to Them.',
      abstract:
        'Millions of older Americans, chronically ill and trapped at home, pose a big challenge to inoculation efforts. Doctors are on it.',
      url:
        'https://www.nytimes.com/2021/03/22/health/virus-seniors-vaccination.html',
      uri: 'nyt://article/e31cffee-6e8f-54eb-95dc-3ab6337fdb18',
      byline: 'By Roni Caryn Rabin and Chang W. Lee',
      item_type: 'Article',
      published_date: '2021-03-22T17:08:27-04:00',
      multimedia: [
        {
          url:
            'https://static01.nyt.com/images/2021/03/22/science/00VIRUS-HOMEBOUND-promo/00VIRUS-HOMEBOUND-promo-articleInline.jpg',
        },
      ],
    },
  ],
};

describe('List Articles Component', () => {
  it('should be contains List Articles text', () => {
    const { UNSAFE_getAllByType } = renderWithTheme(
      <ListArticles
        data={mock}
        loading={false}
        error={false}
        onRefresh={jest.fn()}
        onCardPress={jest.fn()}
      />,
    );
    expect(UNSAFE_getAllByType(ListArticles).length).toBe(1);
  });
});
