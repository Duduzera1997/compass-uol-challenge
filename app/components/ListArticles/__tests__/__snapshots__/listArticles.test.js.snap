// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Articles Component should be contains List Articles in the screen 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "paddingBottom": 20,
      "paddingLeft": 20,
      "paddingRight": 20,
    }
  }
  data={
    Array [
      Object {
        "abstract": "Millions of older Americans, chronically ill and trapped at home, pose a big challenge to inoculation efforts. Doctors are on it.",
        "byline": "By Roni Caryn Rabin and Chang W. Lee",
        "item_type": "Article",
        "multimedia": Array [
          Object {
            "url": "https://static01.nyt.com/images/2021/03/22/science/00VIRUS-HOMEBOUND-promo/00VIRUS-HOMEBOUND-promo-articleInline.jpg",
          },
        ],
        "published_date": "2021-03-22T17:08:27-04:00",
        "title": "How to Vaccinate Homebound Seniors? Take the Shots to Them.",
        "uri": "nyt://article/e31cffee-6e8f-54eb-95dc-3ab6337fdb18",
        "url": "https://www.nytimes.com/2021/03/22/health/virus-seniors-vaccination.html",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.15}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[MockFunction]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[MockFunction]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
  testID="component-flatlist"
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 120,
              "paddingBottom": 10,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 9,
              "flexDirection": "row",
              "height": "100%",
              "opacity": 1,
            }
          }
        >
          <View
            flex={1}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "flex-start",
                  "paddingBottom": 2,
                  "paddingLeft": 7,
                  "paddingRight": 5,
                  "paddingTop": 5,
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://static01.nyt.com/images/2021/03/22/science/00VIRUS-HOMEBOUND-promo/00VIRUS-HOMEBOUND-promo-articleInline.jpg",
                }
              }
              style={
                Array [
                  Object {
                    "borderRadius": 10,
                    "height": "95%",
                    "width": "100%",
                  },
                ]
              }
            />
          </View>
          <View
            flex={2.5}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexBasis": 0,
                  "flexGrow": 2.5,
                  "flexShrink": 1,
                  "justifyContent": "flex-start",
                  "paddingBottom": 2,
                  "paddingLeft": 7,
                  "paddingRight": 5,
                  "paddingTop": 5,
                },
              ]
            }
          >
            <Text
              fontBold={true}
              numberOfLines={2}
              size={15}
              style={
                Array [
                  Object {
                    "color": "#333333",
                    "fontFamily": "PTSansNarrowBold",
                    "fontSize": 15,
                    "paddingBottom": 2,
                    "paddingLeft": 4,
                    "paddingRight": 2,
                    "paddingTop": 4,
                    "width": "99%",
                  },
                ]
              }
            >
              How to Vaccinate Homebound Seniors? Take the Shots to Them.
            </Text>
            <Text
              fontBold={true}
              numberOfLines={2}
              size={11.5}
              style={
                Array [
                  Object {
                    "color": "#767676",
                    "fontFamily": "PTSansNarrowBold",
                    "fontSize": 11.5,
                    "paddingBottom": 2,
                    "paddingLeft": 4,
                    "paddingRight": 2,
                    "paddingTop": 0,
                    "width": "99%",
                  },
                ]
              }
            >
              22 de mar√ßo de 2021
            </Text>
            <Text
              numberOfLines={2}
              size={13}
              style={
                Array [
                  Object {
                    "color": "#333333",
                    "fontFamily": "PTSansNarrowRegular",
                    "fontSize": 13,
                    "paddingBottom": 2,
                    "paddingLeft": 4,
                    "paddingRight": 2,
                    "paddingTop": 4,
                    "width": "99%",
                  },
                ]
              }
            >
              Millions of older Americans, chronically ill and trapped at home, pose a big challenge to inoculation efforts. Doctors are on it.
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
